File: autocontext/src/autocontext/investigation/browser_context.py
"""Browser-backed investigation context capture."""

from __future__ import annotations

from dataclasses import dataclass

from autocontext.config.settings import AppSettings
from autocontext.integrations.browser.context_capture import (
    CapturedBrowserContext,
    capture_browser_context,
    render_captured_browser_context,
)


@dataclass(frozen=True, slots=True)
class InvestigationBrowserContext:
    """Stable browser-derived context folded into investigation inputs."""

    url: str
    title: str
    visible_text: str
    html_path: str | None = None
    screenshot_path: str | None = None


def render_investigation_browser_context(context: InvestigationBrowserContext) -> str:
    """Render browser context into prompt-friendly text."""
    return render_captured_browser_context(
        CapturedBrowserContext(
            url=context.url,
            title=context.title,
            visible_text=context.visible_text,
            html_path=context.html_path,
            screenshot_path=context.screenshot_path,
        )
    )


def build_browser_evidence_summary(context: InvestigationBrowserContext) -> str:
    """Condense browser context into a single evidence summary."""
    if context.title and context.visible_text:
        return f"{context.title}\n{context.visible_text}"
    return context.title or context.visible_text or context.url


def capture_investigation_browser_context(
    settings: AppSettings,
    *,
    browser_url: str,
    investigation_name: str,
) -> InvestigationBrowserContext:
    """Capture a single browser snapshot for an investigation."""
    context = capture_browser_context(
        settings,
        browser_url=browser_url,
        evidence_root=(settings.knowledge_root / "_investigations" / investigation_name),
    )
    return InvestigationBrowserContext(
        url=context.url,
        title=context.title,
        visible_text=context.visible_text,
        html_path=context.html_path,
        screenshot_path=context.screenshot_path,
    )


__all__ = [
    "InvestigationBrowserContext",
    "build_browser_evidence_summary",
    "capture_investigation_browser_context",
    "render_investigation_browser_context",
]

File: ts/src/investigation/browser-context.ts
import { join, resolve } from "node:path";

import {
  captureBrowserContextFromUrl,
  renderCapturedBrowserContext,
  type BrowserContextCaptureSettingsLike,
  type CapturedBrowserContext,
} from "../integrations/browser/context-capture.js";
import type { Evidence } from "./investigation-contracts.js";

export interface InvestigationBrowserContext extends CapturedBrowserContext {}

export interface InvestigationBrowserContextSettingsLike extends BrowserContextCaptureSettingsLike {
  readonly knowledgeRoot: string;
}

export interface CaptureInvestigationBrowserContextRequest {
  readonly settings: InvestigationBrowserContextSettingsLike;
  readonly browserUrl: string;
  readonly investigationName: string;
}

export interface InvestigationBrowserContextDependencies {
  readonly captureBrowserContextFromUrl: typeof captureBrowserContextFromUrl;
}

const DEFAULT_DEPENDENCIES: InvestigationBrowserContextDependencies = {
  captureBrowserContextFromUrl,
};

export async function captureInvestigationBrowserContext(
  opts: CaptureInvestigationBrowserContextRequest,
  dependencies: InvestigationBrowserContextDependencies = DEFAULT_DEPENDENCIES,
): Promise<InvestigationBrowserContext> {
  return dependencies.captureBrowserContextFromUrl({
    settings: opts.settings,
    browserUrl: opts.browserUrl,
    evidenceRoot: join(resolve(opts.settings.knowledgeRoot), "_investigations", opts.investigationName),
  });
}

export function renderInvestigationBrowserContext(context: InvestigationBrowserContext): string {
  return renderCapturedBrowserContext(context);
}

export function buildInvestigationBrowserEvidence(context: InvestigationBrowserContext): Evidence {
  return {
    id: "browser_snapshot",
    kind: "browser_snapshot",
    source: context.url,
    summary: buildInvestigationBrowserSummary(context),
    supports: [],
    contradicts: [],
    isRedHerring: false,
  };
}

export function buildInvestigationBrowserSummary(context: InvestigationBrowserContext): string {
  if (context.title && context.visibleText) {
    return `${context.title}\n${context.visibleText}`;
  }
  return context.title || context.visibleText || context.url;
}
